std/deps.ln

```diff
@@ -1,34 +1,331 @@
+    block: emptyBlock,
+    fullBlock: emptyFullBlock,
+    group: group,
+    name: repo,
+    rev: rev,
+    remote: remote,
+  };
+  return new PackageDependency {
+    pkg: pkg,
+    dep: dep,
+  };
+}
+
+export fn add(pkgd: PackageDependency): RawPackage {
+  pkgd.pkg.dependencies.push(pkgd.dep);
+  return pkgd.pkg;
+}
+  
+export fn block(pkg: RawPackage, block: string): RawPackage {
+  pkg.block.push(block);
+  return pkg;
+}
+
+export fn block(pkg: RawPackage, block: Array<string>): RawPackage {
+  pkg.block = pkg.block + block;
+  return pkg;
+}
+
+export fn block(pkgd: PackageDependency, block: string): PackageDependency {
+  pkgd.dep.block.push(block);
+  return pkgd;
+}
+
+export fn block(pkgd: PackageDependency, block: Array<string>): PackageDependency {
+  pkgd.dep.block = pkgd.dep.block + block;
+  return pkgd;
+}
+
+export fn fullBlock(pkg: RawPackage, block: string): RawPackage {
+  pkg.fullBlock.push(block);
+  return pkg;
+}
+
+export fn fullBlock(pkg: RawPackage, block: Array<string>): RawPackage {
+  pkg.fullBlock = pkg.fullBlock + block;
+  return pkg;
+}
+
+export fn fullBlock(pkgd: PackageDependency, block: string): PackageDependency {
+  pkgd.dep.fullBlock.push(block);
+  return pkgd;
+}
+
+export fn fullBlock(pkgd: PackageDependency, block: Array<string>): PackageDependency {
+  pkgd.dep.fullBlock = pkgd.dep.fullBlock + block;
+  return pkgd;
+}
+
+// The `commit` function takes the packge as argument.
+// It install all dependencies added and apply necessary blocks.
+export fn commit(pkg: RawPackage) {
+  // Remove std dep from blacklist
+  const blacklist = getBlacklist(pkg.std);
+  // Create mocks for blacklisted std libs
+  blockStd(blacklist);
+  // Download and install each dep
+  downloadDependencies(pkg.dependencies);
+  // Apply blocks defined for each dependency
+  applyDependencyBlocks(pkg.dependencies);
+  // Block lib for all dependencies once.
+  applyBlock(pkg.block, false, none(), none());
+  // Block lib for all dependencies once. If any dependency is mocking it already remove that behaviour.
+  applyBlock(pkg.fullBlock, true, none(), none());
   emit exit 0;
 }
 
+fn getBlacklist(std: Array<string>): Array<string> {
+  // TODO: add @std/deps  to this list and handle it on bdd tests
+  const stdList = [
+    '@std/app', '@std/avmdaemon', '@std/cmd', '@std/datastore', '@std/http', 
+    '@std/httpcommon', '@std/httpserver', '@std/json', '@std/seq', '@std/tcp',
+    '@std/tcpserver', '@std/trig',
+  ];
+  return stdList.filter(fn (a: string) = !(pkg.std.find(fn (b: string) = eq(a,b)).isOk()));
+}
+
+fn blockStd(blacklist: Array<string>) {
+  const dependencies = 'dependencies';
+  if !directoryExists(dependencies) {
+    makeDirectory(dependencies);
+  }
+  const modulesPath = [dependencies, 'modules'].join('/');
+  if !directoryExists(modulesPath) {
+    makeDirectory(modulesPath);
+  }
+  const stdPath = [modulesPath, 'std'].join('/');
+  if !directoryExists(stdPath) {
+    makeDirectory(stdPath);
+  }
```

Comment:
This pattern is used enough and looks so much like `mkdir -p` that maybe it should be refactored to something like:

```
fn mkdirp(dir: string) {
  if !directoryExists(dir) {
    makeDirectory(dir);
  }
}

...

mkdirp('dependencies');
mkdirp('dependencies/modules');
mkdirp('dependencies/modules/std');
```

Reply:
definitely! forgot to refactor it